src/binary/merkle_tree.rs

```diff
@@ -135,20 +141,123 @@ where
+    /// ```
+    ///
+    /// We observe that the tree has three peaks at positions `03`, `09`, and
+    /// `12`. These peak positions are recorded in the order that they appear,
+    /// reading left to right in the tree structure, and only descend in height.
+    /// These peak positions communicate everything needed to determine the
+    /// remaining internal nodes building upwards to the root position:
+    ///
+    /// ```text
+    ///               07
+    ///              /  \
+    ///             /    \
+    ///            /      \
+    ///           /        \
+    ///          /          \
+    ///         /            \
+    ///       03              11
+    ///      /  \            /  \
+    ///    ...  ...         /    \
+    ///                   09      \
+    ///                  /  \      \
+    ///                ...  ...    12
+    ///
+    /// ```
+    ///
+    /// No additional intermediate nodes or leaves are required to calculate
+    /// the root position.
+    ///
+    /// The positions of the MMR peaks can be deterministically calculated as a
+    /// function of `n + 1` where `n` is the number of leaves in the tree. By
+    /// appending an additional leaf node to the tree, we generate a new tree
+    /// structure with additional internal nodes (N.B.: this may also change the
+    /// root position if the tree is already balanced).
+    ///
+    /// In our example, we add an additional leaf at leaf index `7` (in-order
+    /// index `14`).
+    ///
+    /// ```text
+    ///               07
+    ///              /  \
+    ///             /    \
+    ///            /      \
+    ///           /        \
+    ///          /          \
+    ///         /            \
+    ///       03              11
+    ///      /  \            /  \
+    ///    ...  ...         /    \
+    ///                   09      13
+    ///                  /  \    /  \
+    ///                ...  ... 12  14
+    /// ```
+    ///
+    /// We observe that the path from the root position to our new leaf position
+    /// yields a set of side positions that includes our original peak
+    /// positions (see [Path Iterator](crate::common::path_iterator::PathIter)):
+    ///
+    /// | Path position | Side position |
+    /// |---------------|---------------|
+    /// |            07 |            07 |
+    /// |            11 |            03 |
+    /// |            13 |            09 |
+    /// |            14 |            12 |
+    ///
+    /// By excluding the root position `07`, we have established the set of
+    /// side positions `03`, `09`, and `12`, matching our set of MMR peaks.
+    ///
     fn build(&mut self) -> Result<(), MerkleTreeError<StorageError>> {
-        let keys = (0..self.leaves_count).map(|i| Position::from_leaf_index(i).in_order_index());
-        for key in keys {
+        // Define a new tree with a leaf count 1 greater than the current leaf
+        // count.
+        let leaves_count = self.leaves_count + 1;
```

Comment:
```suggestion
        let leaves_count = self.leaves_count .saturating_add(1);
```
Or maybe even `.checked_add(1)` if we want to know bout the overflow.

Reply:
Saturating the add will alter the algorithm and yield unexpected results. I think in the case of integer overflow, we should return an error instead. Since `load` is a fallible operation already, it seems reasonable to simply return `Err` if we ever reach the maximum of `u64`. If that is a concern, I can used `checked_add` for that.